src/core/ext/transport/chaotic_good/client_transport.h

```diff
@@ -55,41 +66,101 @@ class ClientTransport {
     if (writer_ != nullptr) {
       writer_.reset();
     }
+    if (reader_ != nullptr) {
+      reader_.reset();
+    }
   }
   auto AddStream(CallArgs call_args) {
     // At this point, the connection is set up.
     // Start sending data frames.
     uint64_t stream_id;
+    // Max buffer is set to 4, so that for stream reads each time it will queue
+    // at most 2 frames.
+    std::shared_ptr<MpscReceiver<ServerFrame>> server_frames =
+        std::make_shared<MpscReceiver<ServerFrame>>(4);
     {
       MutexLock lock(&mu_);
       stream_id = next_stream_id_++;
+      stream_map_.insert(
+          std::pair<uint32_t, std::shared_ptr<MpscSender<ServerFrame>>>(
+              stream_id, std::make_shared<MpscSender<ServerFrame>>(
+                             server_frames->MakeSender())));
     }
-    return Seq(
-        // Continuously send data frame with client to server messages.
-        ForEach(std::move(*call_args.client_to_server_messages),
-                [stream_id, initial_frame = true,
-                 client_initial_metadata =
-                     std::move(call_args.client_initial_metadata),
-                 outgoing_frames = outgoing_frames_.MakeSender()](
-                    MessageHandle result) mutable {
-                  ClientFragmentFrame frame;
-                  frame.stream_id = stream_id;
-                  frame.message = std::move(result);
-                  if (initial_frame) {
-                    // Send initial frame with client intial metadata.
-                    frame.headers = std::move(client_initial_metadata);
-                    initial_frame = false;
-                  }
-                  return Seq(
-                      outgoing_frames.Send(ClientFrame(std::move(frame))),
-                      [](bool success) -> absl::Status {
-                        if (!success) {
-                          return absl::InternalError(
-                              "Send frame to outgoing_frames failed.");
-                        }
-                        return absl::OkStatus();
-                      });
-                }));
+    return TrySeq(
+        TryJoin(
+            // Continuously send client frame with client to server messages.
+            ForEach(std::move(*call_args.client_to_server_messages),
+                    [stream_id, initial_frame = true,
+                     client_initial_metadata =
+                         std::move(call_args.client_initial_metadata),
+                     outgoing_frames = outgoing_frames_.MakeSender()](
+                        MessageHandle result) mutable {
+                      ClientFragmentFrame frame;
+                      frame.stream_id = stream_id;
+                      frame.message = std::move(result);
+                      if (initial_frame) {
```

Comment:
I may have missed this in the previous review. According to the chaotic good specification, Isn't the first frame sent and received by the client expected to be Settings frame instead of a ClientFragmentFrame and a ServerFragmentFrame respectively ?

Reply:
The setting frames are sent during the connection setup path (internal link: https://docs.google.com/document/d/1jQFfG1TGc6VUq54L0apQv2aP5JRbsVe_wMl06m4xG9w/edit?resourcekey=0-6ZkXEgvGvGKnQLpezGgDuw#bookmark=id.qe7uy5kftwqo), which is finished before the `AddStream` (request/response path). We will work on the connection setup step in following PRs.